Fix `try-runtime follow-chain`, add pre/post runtime upgrade tuple test

liamaharon opened the pull request:
- Closes https://github.com/paritytech/substrate/issues/13696
- Adds test validating that `pre_upgrade` and `post_upgrade` hooks are executed correctly for tuple migrations (related: https://github.com/paritytech/substrate/issues/13681)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2628250

kianenigma commented on frame/support/src/traits/hooks.rs:
```diff
@@ -359,10 +359,64 @@ pub trait OnTimestampSet<Moment> {
+
+		TestExternalities::default().execute_with(|| {
+			Foo::try_on_runtime_upgrade(true).unwrap();
+			// todo unify the API for storage parameter_types! output.
```
I seem to have had a leftover TODO here which you can also remove and optionally pursue in a separate PR. 

Here's what is going on. 

The `parameter_types` works in 3 modes. 

1. static (generates thread_local stuff)
2. storage 
3. const 

All of which generate a `Get` implementation, backed by something, along some other helper functions like `get`, `set`, `mutate` and `take` (kinda similar to storage items..). 

It seemed to me that the set of helper functions that is generated by `parameter_types!` is not uniform across different modes. 

If interested, you can convert this comment into an issue and self-assign.

liamaharon replied:
Sounds good, created https://github.com/paritytech/substrate/issues/13795

kianenigma reviewed (commented):
New tests to prevent this happening again?

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2628859

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641019

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641007

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641017

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641008

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641006

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641002

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641314

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641315

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641367

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641366

liamaharon commented:
Hey @kianenigma, I've added a test for `follow-chain`.

The test 
1. Spins up a `--dev` node child process and waits for it to validate it's first block
2. Spins up a `try-runtime follow-chain` child process and waits for it to successfully follow at least 3 blocks

Misc extra notes/questions about the changes:
- I moved cli testing utils from `bin/node/cli/tests/common.rs` to `test-utils/cli/src/lib.rs` so they could be shared between node CLI tests and try-runtime CLI tests, and added a new util in there.
- The existing `Command::new(cargo_bin("substrate"))` pattern used to spawn a `substrate` child process seems to use `target/debug/substrate`. This means that a test could be broken by some changes, but the tests would pass if the project was not re-built. This feels a bit off, I'm wondering if there could be a better approach, to make sure the latest code is always used?

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641719

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641718

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641720

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641714

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641725

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641729

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641770

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641771

liamaharon commented:
Unsure how to fix the `cargo-check-benches` job.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641847

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2641848

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate-macos
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2642103

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2642101

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2642102

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2642091

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2642092

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2642236

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2644226

liamaharon commented:
> The existing Command::new(cargo_bin("substrate")) pattern used to spawn a substrate child process seems to use target/debug/substrate. This means that a test could be broken by some changes, but the tests would pass if the project was not re-built. This feels a bit off, I'm wondering if there could be a better approach, to make sure the latest code is always used?

I found the solution to this. Converted back into a draft as I fix it.

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2645312

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2645901

kianenigma commented on utils/frame/try-runtime/cli/tests/follow_chain.rs:
```diff
@@ -0,0 +1,65 @@
+	use tokio::process::{Child, Command};
+
+	#[tokio::test]
+	async fn follow_chain_works() {
```
now that we have the infra, should we make tests for all commands? can be an issue that you either do yourself, or mentor.

koute replied:
Yes, I think we should have tests for subcommands too; otherwise they'll just accidentally break sooner or later.

liamaharon replied:
Absolutely, it's tracked in this issue: https://github.com/paritytech/substrate/issues/13796. I wanted to test just one command and get the approach reviewed before adding tests for the rest of them. 

There're issues reported with some of the other `try-runtime` commands that I'm close to getting started on, I'll utilise this test infra while debugging those and leave the tests in the fix PR.

kianenigma reviewed (approved):
LGTM, but would be good if someone from @paritytech/sdk-node also takes a look 👍

koute commented on test-utils/cli/src/lib.rs:
```diff
@@ -26,12 +26,146 @@ use nix::{
+		.parent()
+		.expect("Failed to find root workspace directory");
+	let output = Command::new("cargo")
+		.arg("b")
```
```suggestion
		.arg("build")
```

Polkadot-Forum commented:
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/april-updates-for-substrate-and-polkadot-devs/2764/1
